### ISSUE TITLE
Initializing a member with an abstract const passes typechecker, doesn't work at runtime

### ISSUE THREAD

**Issue by fredemmott** [facebook/hhvm#6018]
``` Hack

<?hh
abstract class A {
  abstract const DEFAULT_NAME;
  protected string $name = DEFAULT_NAME;
}
class B extends A {
  const string DEFAULT_NAME = 'hello';
  public function __construct() {
    var_dump($this->name);
  }
}
new B();
```

```
Notice: Use of undefined constant DEFAULT_NAME - assumed 'DEFAULT_NAME' in /home/fred/hhtest/test.php on line 5
```

**Comment by fredemmott** [comment]
The typechecker also likes ` = static::DEFAULT_NAME`, but the runtime also doesn't like it

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

This issue is fixed in hackC 4.56.0, but I can't track down the version in which this was fixed.

Class A initialized `$a->name` is the global constant `\DEFAULT_NAME`, instead of the class constant in your repro.

Switching to `static::DEFAULT_NAME` results in this error and matches runtime behavior.
```
Parsing[1002] Cannot use static or parent::class in property declaration
   --> file.hack
  9 |     protected string $name = static::DEFAULT_NAME;
    |                            ^^^^^^^^^^^^^^^^^^^^^

1 error found.
```

Switching to `self::DEFAULT_NAME` emits this error. The runtime error states that there is no constant `A::DEFAULT_NAME`.
_Fatal error: Uncaught Error: Couldn't find constant A::DEFAULT_NAME_.
That is actually fine, I guess, because you could never observe its value directly, you must observe it value via a subclass.
```
Typing[4129] Cannot reference abstract constant A::DEFAULT_NAME directly
   --> file.hack
  8 |     abstract const string DEFAULT_NAME;
    |                           ^^^^^^^^^^^^ Declaration is here
  9 |     protected string $name = self::DEFAULT_NAME;
    |                                    ^^^^^^^^^^^^

1 error found.
```

It is not possible to write a default static initializer for `A::$name` with the late bound `DEFAULT_NAME`. You must use `B::DEFAULT_NAME`.
